File: backend/app/core/sentry.py
"""
Sentry integration for error tracking and performance monitoring
"""

import sentry_sdk
from sentry_sdk.integrations.fastapi import FastApiIntegration
from sentry_sdk.integrations.sqlalchemy import SqlalchemyIntegration
from sentry_sdk.integrations.redis import RedisIntegration
from sentry_sdk.integrations.logging import LoggingIntegration
import logging
import os
from typing import Optional

from app.core.config import settings


def init_sentry():
    """
    Initialize Sentry SDK for error tracking and performance monitoring
    """
    sentry_dsn = os.getenv("SENTRY_DSN")

    if not sentry_dsn:
        logging.info("Sentry DSN not configured. Skipping Sentry initialization.")
        return

    # Determine environment
    environment = os.getenv("ENVIRONMENT", "development")

    # Initialize Sentry
    sentry_sdk.init(
        dsn=sentry_dsn,
        # Set environment (production, staging, development)
        environment=environment,

        # Release version (use git commit hash or version number)
        release=os.getenv("GIT_COMMIT_SHA", "unknown"),

        # Enable integrations
        integrations=[
            # FastAPI integration for request/response tracking
            FastApiIntegration(
                transaction_style="url",  # Group transactions by URL pattern
            ),

            # SQLAlchemy integration for database query tracking
            SqlalchemyIntegration(),

            # Redis integration for cache tracking
            RedisIntegration(),

            # Logging integration
            LoggingIntegration(
                level=logging.INFO,  # Capture info and above as breadcrumbs
                event_level=logging.ERROR,  # Send errors and above as events
            ),
        ],

        # Performance monitoring
        traces_sample_rate=get_traces_sample_rate(environment),

        # Profiling (CPU and memory usage)
        profiles_sample_rate=get_profiles_sample_rate(environment),

        # Error sampling
        sample_rate=1.0,  # Capture 100% of errors

        # Maximum breadcrumbs
        max_breadcrumbs=50,

        # Attach stack traces to logs
        attach_stacktrace=True,

        # Send default PII (Personally Identifiable Information)
        send_default_pii=False,  # Disable for privacy

        # Debug mode (only in development)
        debug=environment == "development",

        # Before send hook for filtering/modifying events
        before_send=before_send_hook,

        # Before send transaction hook
        before_send_transaction=before_send_transaction_hook,
    )

    logging.info(f"Sentry initialized for environment: {environment}")


def get_traces_sample_rate(environment: str) -> float:
    """
    Get performance tracing sample rate based on environment

    Args:
        environment: Current environment (production, staging, development)

    Returns:
        Sample rate (0.0 to 1.0)
    """
    rates = {
        "production": 0.1,  # Sample 10% of transactions in production
        "staging": 0.5,     # Sample 50% in staging
        "development": 1.0,  # Sample 100% in development
    }
    return rates.get(environment, 0.1)


def get_profiles_sample_rate(environment: str) -> float:
    """
    Get profiling sample rate based on environment

    Args:
        environment: Current environment (production, staging, development)

    Returns:
        Sample rate (0.0 to 1.0)
    """
    rates = {
        "production": 0.01,  # Profile 1% of transactions in production
        "staging": 0.1,      # Profile 10% in staging
        "development": 0.5,   # Profile 50% in development
    }
    return rates.get(environment, 0.01)


def before_send_hook(event, hint):
    """
    Hook called before sending events to Sentry
    Use this to filter or modify events

    Args:
        event: Sentry event dictionary
        hint: Additional context about the event

    Returns:
        Modified event or None to drop the event
    """
    # Filter out specific exceptions if needed
    if "exc_info" in hint:
        exc_type, exc_value, tb = hint["exc_info"]

        # Don't send 404 errors
        if "404" in str(exc_value):
            return None

        # Don't send validation errors (they're expected)
        if "ValidationError" in str(exc_type):
            return None

    # Remove sensitive data from event
    if "request" in event:
        # Remove authorization headers
        if "headers" in event["request"]:
            if "Authorization" in event["request"]["headers"]:
                event["request"]["headers"]["Authorization"] = "[Filtered]"
            if "X-API-Key" in event["request"]["headers"]:
                event["request"]["headers"]["X-API-Key"] = "[Filtered]"

        # Remove sensitive query parameters
        if "query_string" in event["request"]:
            sensitive_params = ["api_key", "token", "password", "secret"]
            for param in sensitive_params:
                if param in event["request"]["query_string"]:
                    event["request"]["query_string"] = event["request"]["query_string"].replace(
                        param, "[Filtered]"
                    )

    return event


def before_send_transaction_hook(event, hint):
    """
    Hook called before sending transactions to Sentry
    Use this to filter or modify transaction events

    Args:
        event: Sentry transaction event
        hint: Additional context

    Returns:
        Modified event or None to drop the event
    """
    # Don't send health check transactions
    if event.get("transaction") == "/health":
        return None

    # Don't send static file transactions
    if event.get("transaction", "").startswith("/static/"):
        return None

    return event


def set_user_context(user_id: str, email: Optional[str] = None, organization_id: Optional[str] = None):
    """
    Set user context for Sentry events

    Args:
        user_id: User ID
        email: User email (optional)
        organization_id: Organization/tenant ID (optional)
    """
    sentry_sdk.set_user({
        "id": user_id,
        "email": email,
        "organization_id": organization_id,
    })


def set_context(key: str, data: dict):
    """
    Set custom context for Sentry events

    Args:
        key: Context key
        data: Context data dictionary
    """
    sentry_sdk.set_context(key, data)


def add_breadcrumb(message: str, category: str = "default", level: str = "info", data: Optional[dict] = None):
    """
    Add a breadcrumb to the current scope

    Args:
        message: Breadcrumb message
        category: Category (e.g., "auth", "db", "api")
        level: Level (debug, info, warning, error, critical)
        data: Additional data dictionary
    """
    sentry_sdk.add_breadcrumb(
        message=message,
        category=category,
        level=level,
        data=data or {},
    )


def capture_exception(exception: Exception, **kwargs):
    """
    Manually capture an exception

    Args:
        exception: Exception to capture
        **kwargs: Additional context
    """
    sentry_sdk.capture_exception(exception, **kwargs)


def capture_message(message: str, level: str = "info", **kwargs):
    """
    Manually capture a message

    Args:
        message: Message to capture
        level: Level (debug, info, warning, error, fatal)
        **kwargs: Additional context
    """
    sentry_sdk.capture_message(message, level=level, **kwargs)


def start_transaction(name: str, op: str):
    """
    Start a new transaction for performance monitoring

    Args:
        name: Transaction name
        op: Operation type (e.g., "http.server", "db.query")

    Returns:
        Transaction context manager
    """
    return sentry_sdk.start_transaction(name=name, op=op)


def start_span(op: str, description: Optional[str] = None):
    """
    Start a new span within a transaction

    Args:
        op: Operation type
        description: Span description

    Returns:
        Span context manager
    """
    return sentry_sdk.start_span(op=op, description=description)


# ============================================================================
# DECORATORS
# ============================================================================

def monitor_performance(op: str = "function"):
    """
    Decorator to monitor function performance

    Args:
        op: Operation type for span

    Example:
        @monitor_performance(op="service.create_deal")
        def create_deal(...):
            ...
    """
    def decorator(func):
        def wrapper(*args, **kwargs):
            with start_span(op=op, description=func.__name__):
                return func(*args, **kwargs)
        return wrapper
    return decorator


async def monitor_async_performance(op: str = "async.function"):
    """
    Decorator to monitor async function performance

    Args:
        op: Operation type for span

    Example:
        @monitor_async_performance(op="service.fetch_data")
        async def fetch_data(...):
            ...
    """
    def decorator(func):
        async def wrapper(*args, **kwargs):
            with start_span(op=op, description=func.__name__):
                return await func(*args, **kwargs)
        return wrapper
    return decorator


# ============================================================================
# USAGE EXAMPLES
# ============================================================================

"""
# In main.py:
from app.core.sentry import init_sentry

@app.on_event("startup")
async def startup():
    init_sentry()

# In route handlers:
from app.core.sentry import set_user_context, add_breadcrumb, capture_exception

@app.get("/api/deals")
async def get_deals(current_user = Depends(get_current_user)):
    # Set user context
    set_user_context(
        user_id=current_user.id,
        email=current_user.email,
        organization_id=current_user.organization_id
    )

    # Add breadcrumb
    add_breadcrumb(
        message="Fetching deals",
        category="api",
        data={"organization_id": current_user.organization_id}
    )

    try:
        # Your code here
        pass
    except Exception as e:
        # Capture exception
        capture_exception(e)
        raise

# Performance monitoring:
from app.core.sentry import start_transaction, start_span

with start_transaction(name="process_deal", op="task"):
    with start_span(op="db.query", description="fetch deal"):
        deal = db.query(Deal).filter_by(id=deal_id).first()

    with start_span(op="service.calculate", description="calculate valuation"):
        valuation = calculate_valuation(deal)

    with start_span(op="db.insert", description="save valuation"):
        db.add(valuation)
        db.commit()
"""

File: frontend/src/lib/sentry.ts
/**
 * Sentry integration for frontend error tracking and performance monitoring
 */

import * as Sentry from '@sentry/react';
import { BrowserTracing } from '@sentry/tracing';

/**
 * Initialize Sentry SDK
 * Call this in your main.tsx or App.tsx before mounting the app
 */
export function initSentry() {
  const sentryDsn = import.meta.env.VITE_SENTRY_DSN;

  if (!sentryDsn) {
    console.info('Sentry DSN not configured. Skipping Sentry initialization.');
    return;
  }

  const environment = import.meta.env.VITE_ENVIRONMENT || 'development';
  const release = import.meta.env.VITE_GIT_COMMIT_SHA || 'unknown';

  Sentry.init({
    dsn: sentryDsn,

    // Set environment
    environment,

    // Set release version
    release,

    // Enable integrations
    integrations: [
      // Browser tracing for performance monitoring
      new BrowserTracing({
        // Set tracing origins to match your API
        tracingOrigins: [
          'localhost',
          /^\//,  // Relative URLs
          /^https:\/\/[^/]*\.render\.com/,  // Render.com APIs
        ],

        // React Router integration
        routingInstrumentation: Sentry.reactRouterV6Instrumentation(
          // Will be available after React Router setup
        ),
      }),

      // Replay integration for session recording
      new Sentry.Replay({
        // Mask all text content for privacy
        maskAllText: true,
        // Block all media (images, videos) for privacy
        blockAllMedia: true,
        // Sample rate for normal sessions
        sessionSampleRate: environment === 'production' ? 0.1 : 1.0,
        // Sample rate for sessions with errors
        errorSampleRate: 1.0,
      }),
    ],

    // Performance monitoring sample rate
    tracesSampleRate: getTracesSampleRate(environment),

    // Error sampling
    sampleRate: 1.0,  // Capture 100% of errors

    // Maximum breadcrumbs
    maxBreadcrumbs: 50,

    // Attach stack traces
    attachStacktrace: true,

    // Don't send default PII
    sendDefaultPII: false,

    // Debug mode (only in development)
    debug: environment === 'development',

    // Before send hook
    beforeSend: beforeSendHook,

    // Before send transaction hook
    beforeSendTransaction: beforeSendTransactionHook,

    // Ignore specific errors
    ignoreErrors: [
      // Browser extension errors
      'top.GLOBALS',
      'chrome-extension://',
      'moz-extension://',

      // Network errors
      'NetworkError',
      'Network request failed',
      'Failed to fetch',

      // Script loading errors
      'ResizeObserver loop limit exceeded',
      'ResizeObserver loop completed with undelivered notifications',

      // Common false positives
      'Non-Error promise rejection captured',
    ],

    // Deny URLs (don't report errors from these sources)
    denyUrls: [
      /extensions\//i,
      /^chrome:\/\//i,
      /^moz-extension:\/\//i,
    ],
  });

  console.info(`Sentry initialized for environment: ${environment}`);
}

/**
 * Get traces sample rate based on environment
 */
function getTracesSampleRate(environment: string): number {
  const rates: Record<string, number> = {
    production: 0.1,   // Sample 10% in production
    staging: 0.5,      // Sample 50% in staging
    development: 1.0,  // Sample 100% in development
  };
  return rates[environment] || 0.1;
}

/**
 * Before send hook - filter/modify events before sending
 */
function beforeSendHook(event: Sentry.Event, hint: Sentry.EventHint): Sentry.Event | null {
  // Filter out specific error types
  if (hint.originalException) {
    const error = hint.originalException;

    // Don't send network timeouts
    if (error instanceof Error && error.message.includes('timeout')) {
      return null;
    }

    // Don't send 404 errors
    if (error instanceof Error && error.message.includes('404')) {
      return null;
    }
  }

  // Remove sensitive data from request
  if (event.request) {
    // Remove authorization headers
    if (event.request.headers) {
      if (event.request.headers['Authorization']) {
        event.request.headers['Authorization'] = '[Filtered]';
      }
      if (event.request.headers['X-API-Key']) {
        event.request.headers['X-API-Key'] = '[Filtered]';
      }
    }

    // Remove sensitive query parameters
    if (event.request.query_string) {
      const sensitiveParams = ['api_key', 'token', 'password', 'secret'];
      sensitiveParams.forEach(param => {
        if (event.request?.query_string?.includes(param)) {
          event.request.query_string = event.request.query_string.replace(
            new RegExp(`${param}=[^&]*`, 'gi'),
            `${param}=[Filtered]`
          );
        }
      });
    }
  }

  return event;
}

/**
 * Before send transaction hook - filter/modify transactions
 */
function beforeSendTransactionHook(
  event: Sentry.Event,
  hint: Sentry.EventHint
): Sentry.Event | null {
  // Don't send transactions for health checks
  if (event.transaction === '/health' || event.transaction === '/api/health') {
    return null;
  }

  // Don't send transactions for static assets
  if (event.transaction?.match(/\.(js|css|png|jpg|svg|woff|woff2)$/)) {
    return null;
  }

  return event;
}

/**
 * Set user context for Sentry
 */
export function setUserContext(
  userId: string,
  email?: string,
  organizationId?: string
) {
  Sentry.setUser({
    id: userId,
    email,
    organization_id: organizationId,
  });
}

/**
 * Clear user context (on logout)
 */
export function clearUserContext() {
  Sentry.setUser(null);
}

/**
 * Set custom context
 */
export function setContext(key: string, data: Record<string, any>) {
  Sentry.setContext(key, data);
}

/**
 * Add a breadcrumb
 */
export function addBreadcrumb(
  message: string,
  category: string = 'default',
  level: Sentry.SeverityLevel = 'info',
  data?: Record<string, any>
) {
  Sentry.addBreadcrumb({
    message,
    category,
    level,
    data,
  });
}

/**
 * Manually capture an exception
 */
export function captureException(error: Error, context?: Record<string, any>) {
  if (context) {
    Sentry.withScope((scope) => {
      Object.entries(context).forEach(([key, value]) => {
        scope.setContext(key, value);
      });
      Sentry.captureException(error);
    });
  } else {
    Sentry.captureException(error);
  }
}

/**
 * Manually capture a message
 */
export function captureMessage(
  message: string,
  level: Sentry.SeverityLevel = 'info',
  context?: Record<string, any>
) {
  if (context) {
    Sentry.withScope((scope) => {
      Object.entries(context).forEach(([key, value]) => {
        scope.setContext(key, value);
      });
      Sentry.captureMessage(message, level);
    });
  } else {
    Sentry.captureMessage(message, level);
  }
}

/**
 * Start a transaction for performance monitoring
 */
export function startTransaction(name: string, op: string) {
  return Sentry.startTransaction({ name, op });
}

/**
 * Create an error boundary component
 * Use this to wrap your app or specific components
 */
export const SentryErrorBoundary = Sentry.ErrorBoundary;

/**
 * Fallback component for error boundary
 */
export function ErrorFallback({ error, resetError }: { error: Error; resetError: () => void }) {
  return (
    <div style={{ padding: '2rem', textAlign: 'center' }}>
      <h1>Something went wrong</h1>
      <p>We've been notified and are looking into it.</p>
      <pre style={{ textAlign: 'left', padding: '1rem', background: '#f5f5f5' }}>
        {error.message}
      </pre>
      <button onClick={resetError} style={{ marginTop: '1rem', padding: '0.5rem 1rem' }}>
        Try again
      </button>
    </div>
  );
}

/**
 * HOC to profile component performance
 */
export const withProfiler = Sentry.withProfiler;

// Export Sentry for direct access if needed
export { Sentry };

/**
 * Usage Example:
 *
 * // In main.tsx:
 * import { initSentry } from './lib/sentry';
 * initSentry();
 *
 * // In App.tsx:
 * import { SentryErrorBoundary, ErrorFallback } from './lib/sentry';
 *
 * function App() {
 *   return (
 *     <SentryErrorBoundary fallback={ErrorFallback}>
 *       <Router>
 *         <Routes>...</Routes>
 *       </Router>
 *     </SentryErrorBoundary>
 *   );
 * }
 *
 * // In auth context:
 * import { setUserContext, clearUserContext } from './lib/sentry';
 *
 * function login(user) {
 *   setUserContext(user.id, user.email, user.organizationId);
 * }
 *
 * function logout() {
 *   clearUserContext();
 * }
 *
 * // In API calls:
 * import { addBreadcrumb, captureException } from './lib/sentry';
 *
 * async function fetchData() {
 *   addBreadcrumb('Fetching deals', 'api', 'info');
 *   try {
 *     const response = await fetch('/api/deals');
 *     return await response.json();
 *   } catch (error) {
 *     captureException(error, { context: 'fetchDeals' });
 *     throw error;
 *   }
 * }
 *
 * // With profiler:
 * import { withProfiler } from './lib/sentry';
 *
 * export default withProfiler(MyExpensiveComponent);
 */
